src/librustc_typeck/collect.rs

```diff
@@ -224,7 +224,6 @@ impl Visitor<'tcx> for CollectItemTypesVisitor<'tcx> {
         if let hir::ExprKind::Closure(..) = expr.kind {
             let def_id = self.tcx.hir().local_def_id(expr.hir_id);
             self.tcx.ensure().generics_of(def_id);
-            self.tcx.ensure().type_of(def_id);
```

Comment:
I removed this line to make 29 UI tests pass - their only failure mode was missing *some* of the errors.

I believe what's happening is that this would now trigger type-checking the body of the closure, before bodies are normally type-checked, and so it trips on an "abort if any errors" check.

Can we start removing those "checkpoints" that turn regular errors into fatal ones, given how on-demand some of the various checks are nowadays? cc @estebank

Reply:
I think there are only a handful of `FatalError.raise()`s left, and the ones that do remain do so because they'd require a much larger refactor to get rid of them.

Reply:
If they're because of having to return a value, I suspect using `Result<T, ErrorReported>` could work well.